Src/Fido2/TrustAnchor.cs

```diff
@@ -1,60 +1,65 @@
-                }
-            }
-
-            else if (ContainsAttestationType(metadataEntry, MetadataAttestationType.ATTESTATION_ANONCA))
-            {
-                // skip verification for Anonymization CA (AnonCA)
-            }
-            else // otherwise, ensure the certificate is self signed
-            {
-                var trustPath0 = trustPath[0];
-
-                if (!string.Equals(trustPath0.Subject, trustPath0.Issuer, StringComparison.Ordinal))
-                {
-                    // TODO: Improve this error message
-                    throw new Fido2VerificationException("Attestation with full attestation from authenticator that does not support full attestation");
-                }
-            }
-
-            // TODO: Verify all MetadataAttestationTypes are correctly handled
-
-            // [ ] ATTESTATION_ECDAA "ecdaa"    | currently handled as self signed  w/ no test coverage
-            // [ ] ATTESTATION_ANONCA "anonca"  | currently not verified            w/ no test coverage
-            // [ ] ATTESTATION_NONE "none"      | currently handled as self signed  w/ no test coverage               
-        }
-    }
-}
+﻿using System;
+using System.Linq;
+using System.Security.Cryptography.X509Certificates;
+
+using Fido2NetLib.Exceptions;
+
+namespace Fido2NetLib;
+
+public static class TrustAnchor
+{
+    public static void Verify(MetadataBLOBPayloadEntry? metadataEntry, X509Certificate2[] trustPath, bool conformance)
+    {
+        if (trustPath != null && metadataEntry?.MetadataStatement?.AttestationTypes is not null)
+        {
+            static bool ContainsAttestationType(MetadataBLOBPayloadEntry entry, MetadataAttestationType type)
+            {
+                return entry.MetadataStatement.AttestationTypes.Contains(type.ToEnumMemberValue());
+            }
+
+            // If the authenticator's metadata requires basic full attestation, build and verify the chain
+            if (ContainsAttestationType(metadataEntry, MetadataAttestationType.ATTESTATION_BASIC_FULL) ||
+                ContainsAttestationType(metadataEntry, MetadataAttestationType.ATTESTATION_PRIVACY_CA))
+            {
+                string[] certStrings = metadataEntry.MetadataStatement.AttestationRootCertificates;
+                var attestationRootCertificates = new X509Certificate2[certStrings.Length];
+
+                for (int i = 0; i < attestationRootCertificates.Length; i++)
+                {
+                    attestationRootCertificates[i] = new X509Certificate2(Convert.FromBase64String(certStrings[i]));
+                }
+
+                if (trustPath.Length > 1 && attestationRootCertificates.Any(c => string.Equals(c.Thumbprint, trustPath[^1].Thumbprint, StringComparison.Ordinal)))
+                {
+                    throw new Fido2VerificationException(Fido2ErrorMessages.InvalidCertificateChain);
+                }

```

Comment:
@aseigler Would appreciate your eyes here too, but to me it just looks like we have the addition of above if-block that is more strict than without it.

Reply:
@abergs Are we able to reference the rule in https://datatracker.ietf.org/doc/html/rfc5280 that this new condition enforces?

Reply:
https://github.com/passwordless-lib/fido2-net-lib/pull/456

TrustAnchor.cs : 32
This exact testcase was fixed with this (conformance tests v1.7.15):

Server-ServerAuthenticatorAttestationResponse-Resp-5 Test server processing "packed" FULL attestation F-10 Send ServerAuthenticatorAttestationResponse with FULL "packed" attestation, with attStmt.x5c containing full chain, and check that server returns an error https://datatracker.ietf.org/doc/html/rfc5280#section-6.1

Reply:
@googyi Thanks for chiming in. What part of the [datatracker.ietf.org/doc/html/rfc5280#section-6.1](https://datatracker.ietf.org/doc/html/rfc5280#section-6.1) is deciding checking the last entry of the trustpath to root certs?

Reply:
I think this 6.1 section was referenced in that testcase, I don't know which part describes this exactly. 
This looks almost like that: 

"When the trust anchor is provided in the form of a self-signed
   certificate, this self-signed certificate is not included as part of
   the prospective certification path."

   But I am not sure that this is a self-signed cert.

Reply:
6.1 also mentions " A certificate MUST NOT appear more than once in a prospective
   certification path." which may be fixed by this?